### ISSUE TITLE
160 Data Science Interview Questions

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6287]
160+ Data Science Interview Questions<br>
<br>
https://ift.tt/3amjNSZ<br>
<br>
<br>
<br>
<div>
<div>
<img src="https://hackernoon.com/drafts/f11gu3yd0.png">
<p>A typical interview process for a data science position includes multiple rounds. Often, one of such rounds covers theoretical concepts, where the goal is to determine if the candidate knows the fundamentals of machine learning.</p>
<p>In this post, I&rsquo;d like to summarize all my interviewing experience &#8202;&mdash;&#8202; from both interviewing and being interviewed &#8202;&mdash;&#8202; and came up with a list of 160+ theoretical data science questions.</p>
<p>This includes the following topics:</p>
<ul>
<li>Linear regression</li>
<li>Validation</li>
<li>Classification and logistic regression</li>
<li>Regularization</li>
<li>Decision trees</li>
<li>Random forest</li>
<li>Gradient boosting trees</li>
<li>Neural networks</li>
<li>Text classification</li>
<li>Clustering</li>
<li>Ranking: search and recommendation</li>
<li>Time series</li>
</ul>
<p>The number of questions in this post might seem overwhelming &#8202;&mdash;&#8202; and it indeed is. Keep in mind that the interview flow is based on what the company needs and what you have worked with, so if you didn&rsquo;t work with models in time series or computer vision, you shouldn&rsquo;t get questions about them.</p>
<p>Important: don&rsquo;t feel discouraged if you don&rsquo;t know the answers to some of the interview questions. This is absolutely fine.</p>
<p>Finally, to make it simpler, I grouped the questions into three categories, based on difficulty:</p>
<ul>
<li>&#128118; easy</li>
<li>&zwj;&#11088;&#65039; medium</li>
<li>&#128640; expert</li>
</ul>
<p>That&rsquo;s, of course, subjective, and it&rsquo;s based only on my personal opinion.</p>
<p>Let&rsquo;s start!</p>
<p>Supervised machine&nbsp;learning</p>
<ul>
<li>What is supervised machine learning? &#128118;</li>
</ul>
<p>Linear regression</p>
<ul>
<li>What is regression? Which models can you use to solve a regression problem? &#128118;</li>
<li>What is linear regression? When do we use it? &#128118;</li>
<li>What&rsquo;s the normal distribution? Why do we care about it? &#128118;</li>
<li>How do we check if a variable follows the normal distribution? &zwj;&#11088;&#65039;</li>
<li>What if we want to build a model for predicting prices? Are prices distributed normally? Do we need to do any pre-processing for prices? &zwj;&#11088;&#65039;</li>
<li>What are the methods for solving linear regression do you know? &zwj;&#11088;&#65039;</li>
<li>What is gradient descent? How does it work? &zwj;&#11088;&#65039;</li>
<li>What is the normal equation? &zwj;&#11088;&#65039;</li>
<li>What is SGD &#8202;&mdash;&#8202; stochastic gradient descent? What&rsquo;s the difference with the usual gradient descent? &zwj;&#11088;&#65039;</li>
<li>Which metrics for evaluating regression models do you know? &#128118;</li>
<li>What are MSE and RMSE? &#128118;</li>
</ul>
<p>Validation</p>
<ul>
<li>What is overfitting? &#128118;</li>
<li>How to validate your models? &#128118;</li>
<li>Why do we need to split our data into three parts: train, validation, and test? &#128118;</li>
<li>Can you explain how cross-validation works? &#128118;</li>
<li>What is K-fold cross-validation? &#128118;</li>
<li>How do we choose K in K-fold cross-validation? What&rsquo;s your favorite K? &#128118;</li>
</ul>
<p>Classification</p>
<ul>
<li>What is classification? Which models would you use to solve a classification problem? &#128118;</li>
<li>What is logistic regression? When do we need to use it? &#128118;</li>
<li>Is logistic regression a linear model? Why? &#128118;</li>
<li>What is sigmoid? What does it do? &#128118;</li>
<li>How do we evaluate classification models? &#128118;</li>
<li>What is accuracy? &#128118;</li>
<li>Is accuracy always a good metric? &#128118;</li>
<li>What is the confusion table? What are the cells in this table? &#128118;</li>
<li>What is precision, recall, and F1-score? &#128118;</li>
<li>Precision-recall trade-off &zwj;&#11088;&#65039;</li>
<li>What is the ROC curve? When to use it? &zwj;&#11088;&#65039;</li>
<li>What is AUC (AU ROC)? When to use it? &zwj;&#11088;&#65039;</li>
<li>How to interpret the AU ROC score? &zwj;&#11088;&#65039;</li>
<li>What is the PR (precision-recall) curve? &zwj;&#11088;&#65039;</li>
<li>What is the area under the PR curve? Is it a useful metric? &zwj;&#11088;&#65039;</li>
<li>In which cases AU PR is better than AU ROC? &zwj;&#11088;&#65039;</li>
<li>What do we do with categorical variables? &zwj;&#11088;&#65039;</li>
<li>Why do we need one-hot encoding? &zwj;&#11088;&#65039;</li>
</ul>
<p>Regularization</p>
<ul>
<li>What happens to our linear regression model if we have three columns in our data: x, y, z &#8202;&mdash;&#8202; and z is a sum of x and y? &zwj;&#11088;&#65039;</li>
<li>What happens to our linear regression model if the column z in the data is a sum of columns x and y and some random noise? &zwj;&#11088;&#65039;</li>
<li>What is regularization? Why do we need it? &#128118;</li>
<li>Which regularization techniques do you know? &zwj;&#11088;&#65039;</li>
<li>What kind of regularization techniques are applicable to linear models? &zwj;&#11088;&#65039;</li>
<li>How does L2 regularization look like in a linear model? &zwj;&#11088;&#65039;</li>
<li>How do we select the right regularization parameters? &#128118;</li>
<li>What&rsquo;s the effect of L2 regularization on the weights of a linear model? &zwj;&#11088;&#65039;</li>
<li>How L1 regularization looks like in a linear model? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s the difference between L2 and L1 regularization? &zwj;&#11088;&#65039;</li>
<li>Can we have both L1 and L2 regularization components in a linear model? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s the interpretation of the bias term in linear models? &zwj;&#11088;&#65039;</li>
<li>How do we interpret weights in linear models? &zwj;&#11088;&#65039;</li>
<li>If a weight for one variable is higher than for another &#8202;&mdash;&#8202; can we say that this variable is more important? &zwj;&#11088;&#65039;</li>
<li>When do we need to perform feature normalization for linear models? When it&rsquo;s okay not to do it? &zwj;&#11088;&#65039;</li>
</ul>
<p>Feature selection</p>
<ul>
<li>What is feature selection? Why do we need it? &#128118;</li>
<li>Is feature selection important for linear models? &zwj;&#11088;&#65039;</li>
<li>Which feature selection techniques do you know? &zwj;&#11088;&#65039;</li>
<li>Can we use L1 regularization for feature selection? &zwj;&#11088;&#65039;</li>
<li>Can we use L2 regularization for feature selection? &zwj;&#11088;&#65039;</li>
</ul>
<p>Decision trees</p>
<ul>
<li>What are the decision trees? &#128118;</li>
<li>How do we train decision trees? &zwj;&#11088;&#65039;</li>
<li>What are the main parameters of the decision tree model? &#128118;</li>
<li>How do we handle categorical variables in decision trees? &zwj;&#11088;&#65039;</li>
<li>What are the benefits of a single decision tree compared to more complex models? &zwj;&#11088;&#65039;</li>
<li>How can we know which features are more important for the decision tree model? &zwj;&#11088;&#65039;</li>
</ul>
<p>Random forest</p>
<ul>
<li>What is random forest? &#128118;</li>
<li>Why do we need randomization in random forest? &zwj;&#11088;&#65039;</li>
<li>What are the main parameters of the random forest model? &zwj;&#11088;&#65039;</li>
<li>How do we select the depth of the trees in random forest? &zwj;&#11088;&#65039;</li>
<li>How do we know how many trees we need in random forest? &zwj;&#11088;&#65039;</li>
<li>Is it easy to parallelize training of a random forest model? How can we do it? &zwj;&#11088;&#65039;</li>
<li>What are the potential problems with many large trees? &zwj;&#11088;&#65039;</li>
<li>What if instead of finding the best split, we randomly select a few splits and just select the best from them. Will it work? &#128640;</li>
<li>What happens when we have correlated features in our data? &zwj;&#11088;&#65039;</li>
</ul>
<p>Gradient boosting</p>
<ul>
<li>What is gradient boosting trees? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s the difference between random forest and gradient boosting? &zwj;&#11088;&#65039;</li>
<li>Is it possible to parallelize training of a gradient boosting model? How to do it? &zwj;&#11088;&#65039;</li>
<li>Feature importance in gradient boosting trees &#8202;&mdash;&#8202; what are possible options? &zwj;&#11088;&#65039;</li>
<li>Are there any differences between continuous and discrete variables when it comes to feature importance of gradient boosting models? &#128640;</li>
<li>What are the main parameters in the gradient boosting model? &zwj;&#11088;&#65039;</li>
<li>How do you approach tuning parameters in XGBoost or LightGBM? &#128640;</li>
<li>How do you select the number of trees in the gradient boosting model? &zwj;&#11088;&#65039;</li>
</ul>
<p>Parameter tuning</p>
<ul>
<li>Which parameter tuning strategies (in general) do you know? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s the difference between grid search parameter tuning strategy and random search? When to use one or another? &zwj;&#11088;&#65039;</li>
</ul>
<p>Neural networks</p>
<ul>
<li>What kind of problems neural nets can solve? &#128118;</li>
<li>How does a usual fully-connected feed-forward neural network work? &zwj;&#11088;&#65039;</li>
<li>Why do we need activation functions? &#128118;</li>
<li>What are the problems with sigmoid as an activation function? &zwj;&#11088;&#65039;</li>
<li>What is ReLU? How is it better than sigmoid or tanh? &zwj;&#11088;&#65039;</li>
<li>How we can initialize the weights of a neural network? &zwj;&#11088;&#65039;</li>
<li>What if we set all the weights of a neural network to 0? &zwj;&#11088;&#65039;</li>
<li>What regularization techniques for neural nets do you know? &zwj;&#11088;&#65039;</li>
<li>What is dropout? Why is it useful? How does it work? &zwj;&#11088;&#65039;</li>
</ul>
<p>Optimization in neural&nbsp;networks</p>
<ul>
<li>What is backpropagation? How does it work? Why do we need it? &zwj;&#11088;&#65039;</li>
<li>Which optimization techniques for training neural nets do you know? &zwj;&#11088;&#65039;</li>
<li>How do we use SGD (stochastic gradient descent) for training a neural net? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s the learning rate? &#128118;</li>
<li>What happens when the learning rate is too large? Too small? &#128118;</li>
<li>How to set the learning rate? &zwj;&#11088;&#65039;</li>
<li>What is Adam? What&rsquo;s the main difference between Adam and SGD? &zwj;&#11088;&#65039;</li>
<li>When would you use Adam and when SGD? &zwj;&#11088;&#65039;</li>
<li>Do we want to have a constant learning rate or we better change it throughout training? &zwj;&#11088;&#65039;</li>
<li>How do we decide when to stop training a neural net? &#128118;</li>
<li>What is model checkpointing? &zwj;&#11088;&#65039;</li>
<li>Can you tell us how you approach the model training process? &zwj;&#11088;&#65039;</li>
</ul>
<p>Neural networks for computer&nbsp;vision</p>
<ul>
<li>How we can use neural nets for computer vision? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s a convolutional layer? &zwj;&#11088;&#65039;</li>
<li>Why do we actually need convolutions? Can&rsquo;t we use fully-connected layers for that? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s pooling in CNN? Why do we need it? &zwj;&#11088;&#65039;</li>
<li>How does max pooling work? Are there other pooling techniques? &zwj;&#11088;&#65039;</li>
<li>Are CNNs resistant to rotations? What happens to the predictions of a CNN if an image is rotated? &#128640;</li>
<li>What are augmentations? Why do we need them? &#128118;What kind of augmentations do you know? &#128118;How to choose which augmentations to use? &zwj;&#11088;&#65039;</li>
<li>What kind of CNN architectures for classification do you know? &#128640;</li>
<li>What is transfer learning? How does it work? &zwj;&#11088;&#65039;</li>
<li>What is object detection? Do you know any architectures for that? &#128640;</li>
<li>What is object segmentation? Do you know any architectures for that? &#128640;</li>
</ul>
<p>Text classification</p>
<ul>
<li>How can we use machine learning for text classification? &zwj;&#11088;&#65039;</li>
<li>What is bag of words? How we can use it for text classification? &zwj;&#11088;&#65039;</li>
<li>What are the advantages and disadvantages of bag of words? &zwj;&#11088;&#65039;</li>
<li>What are N-grams? How can we use them? &zwj;&#11088;&#65039;</li>
<li>How large should be N for our bag of words when using N-grams? &zwj;&#11088;&#65039;</li>
<li>What is TF-IDF? How is it useful for text classification? &zwj;&#11088;&#65039;</li>
<li>Which model would you use for text classification with bag of words features? &zwj;&#11088;&#65039;</li>
<li>Would you prefer gradient boosting trees model or logistic regression when doing text classification with bag of words? &zwj;&#11088;&#65039;</li>
<li>What are word embeddings? Why are they useful? Do you know Word2Vec? &zwj;&#11088;&#65039;</li>
<li>Do you know any other ways to get word embeddings? &#128640;</li>
<li>If you have a sentence with multiple words, you may need to combine multiple word embeddings into one. How would you do it? &zwj;&#11088;&#65039;</li>
<li>Would you prefer gradient boosting trees model or logistic regression when doing text classification with embeddings? &zwj;&#11088;&#65039;</li>
<li>How can you use neural nets for text classification? &#128640;</li>
<li>How can we use CNN for text classification? &#128640;</li>
</ul>
<p>Clustering</p>
<ul>
<li>What is unsupervised learning? &#128118;</li>
<li>What is clustering? When do we need it? &#128118;</li>
<li>Do you know how K-means works? &zwj;&#11088;&#65039;</li>
<li>How to select K for K-means? &zwj;&#11088;&#65039;</li>
<li>What are the other clustering algorithms do you know? &zwj;&#11088;&#65039;</li>
<li>Do you know how DBScan works? &zwj;&#11088;&#65039;</li>
<li>When would you choose K-means and when DBScan? &zwj;&#11088;&#65039;</li>
</ul>
<p>Dimensionality reduction</p>
<ul>
<li>What is the curse of dimensionality? Why do we care about it? &zwj;&#11088;&#65039;</li>
<li>Do you know any dimensionality reduction techniques? &zwj;&#11088;&#65039;</li>
<li>What&rsquo;s singular value decomposition? How is it typically used for machine learning? &zwj;&#11088;&#65039;</li>
</ul>
<p>Ranking and&nbsp;search</p>
<ul>
<li>What is the ranking problem? Which models can you use to solve them? &zwj;&#11088;&#65039;</li>
<li>What are good unsupervised baselines for text information retrieval? &zwj;&#11088;&#65039;</li>
<li>How would you evaluate your ranking algorithms? Which offline metrics would you use? &zwj;&#11088;&#65039;</li>
<li>What is precision and recall at k? &zwj;&#11088;&#65039;</li>
<li>What is mean average precision at k? &zwj;&#11088;&#65039;</li>
<li>How can we use machine learning for search? &zwj;&#11088;&#65039;</li>
<li>How can we get training data for our ranking algorithms? &zwj;&#11088;&#65039;</li>
<li>Can we formulate the search problem as a classification problem? How? &zwj;&#11088;&#65039;</li>
<li>How can we use clicks data as the training data for ranking algorithms? &#128640;</li>
<li>Do you know how to use gradient boosting trees for ranking? &#128640;</li>
<li>How do you do an online evaluation of a new ranking algorithm? &zwj;&#11088;&#65039;</li>
</ul>
<p>Recommender systems</p>
<ul>
<li>What is a recommender system? &#128118;</li>
<li>What are good baselines when building a recommender system? &zwj;&#11088;&#65039;</li>
<li>What is collaborative filtering? &zwj;&#11088;&#65039;</li>
<li>How we can incorporate implicit feedback (clicks, etc) into our recommender systems? &zwj;&#11088;&#65039;</li>
<li>What is the cold start problem? &zwj;&#11088;&#65039;</li>
<li>Possible approaches to solving the cold start problem? &zwj;&#11088;&#65039;&#128640;</li>
</ul>
<p>Time series</p>
<ul>
<li>What is a time series? &#128118;</li>
<li>How is time series different from the usual regression problem? &#128118;</li>
<li>Which models do you know for solving time series problems? &zwj;&#11088;&#65039;</li>
<li>If there&rsquo;s a trend in our series, how we can remove it? And why would we want to do it? &zwj;&#11088;&#65039;</li>
<li>You have a series with only one variable &ldquo;y&rdquo; measured at time t. How do predict &ldquo;y&rdquo; at time t+1? Which approaches would you use? &zwj;&#11088;&#65039;</li>
<li>You have a series with a variable &ldquo;y&rdquo; and a set of features. How do you predict &ldquo;y&rdquo; at t+1? Which approaches would you use? &zwj;&#11088;&#65039;</li>
<li>What are the problems with using trees for solving time series problems? &zwj;&#11088;&#65039;</li>
</ul>
<p>That was a long list! I hope you found it useful. Good luck with your interviews!</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Hacker Noon<br>
<br>
March 2, 2020 at 04:03PM
